### PR TITLE
Add an "Execution environment" section.

### PR DESCRIPTION
**Pull request by andreimatei**
I've struggled a bit with how to describe the way we're going to schedule parts of the physical plan on nodes. There are some things we want to run in parallel (unrelated parts of a query), there are some things we want to schedule sequentially (many TableReaders). A related question is who decides what TableReaders should run in parallel (e.g. one thread per storage device). I'm thinking of pushing this decision to the guy that makes the physical plan, rather than the local node. See what you think.

@RaduBerinde @knz

---
This change is [Reviewable](https://reviewable.io/reviews/raduberinde/cockroach/7)

### CODE REVIEW

**Comment by RaduBerinde** [comment]
This makes sense, I liked it.

I wouldn't worry too much about the details of parallel execution for TableReaders for the same table yet. As long as the framework doesn't make that very difficult, we can probably look at it later. Though I think the node should be able to decide that (as a scheduling decision), not the physical planner.

We can even consider starting with a single TableReader per table for each node (single flow). For small queries (where it doesn't span ranges) this would be the same, for large queries we would no longer be able to use the inter-stream ordering but otherwise things would still work. We can then look at extending the infrastructure to consider them separate flows. I'm just thinking out loud how to start building things so we have something that works before too long.

I don't know if we should fix the details of how processes map to goroutines at this stage. If we have a sequence of processors in a flow, the synchronization costs of having a goroutine for each one might be too large - it might be much cheaper to do the processing for that flow in one goroutine (like we do it today). Maybe that's what you meant by "fusing".

One thing that will be tricky is index-joins (or in general joins-by-lookups). As we read PK values from the index, we would want each PK query to be directed to the node that stores that row, so that further processing happens on that node. But on the other hand, we don't want to set up flows on all the possible nodes that might have a row for that table (we may be doing a very small number of lookups, or the lookups may turn out to be somewhat localized). So we would need a way to be able to set up the flows "on the fly". We could circumvent this in the beginning by setting up the flows if there is a small number of possible nodes, and otherwise just do all the PK queries on the nodes from where we read the index values.

**Comment by andreimatei** [comment]
PTAL last commit

File: docs/RFCS/distributed_sql.md [hunk 1]
```diff
@@ -361,11 +344,34 @@ to `ScheduleFlows`). This call might arrive to a node even before the
 corresponding `ScheduleFlows` arrives. In this case, the mailbox is created on
 the fly, in the hope that the `ScheduleFlows` will follow soon. If that doesn't
 happen within a timeout, the mailbox is retired.
-Mailboxes present a channel interface to the local processors.
+Mailboxes present a channel interface to the local processors.  
+If we move to a multiple `TableReader`s/flows per node, then we'd probably
+still want one single mailbox output mailbox for all the homogeneous flows (if
```
Comment (RaduBerinde) [inline docs/RFCS/distributed_sql.md]: single output mailbox

Comment (andreimatei) [inline docs/RFCS/distributed_sql.md]: done

**Comment by RaduBerinde** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[docs/RFCS/distributed_sql.md, line 692 \[r3\]](https://reviewable.io:443/reviews/raduberinde/cockroach/7#-KFFaQvo-PJ0kbw8ACrj:-KFFaQvo-PJ0kbw8ACrk:-1307876678) ([raw file](https://github.com/raduberinde/cockroach/blob/d70937c07c782e58cbaeb8b0b059ca6baeb98c07/docs/RFCS/distributed_sql.md#L692)):*
remove probably

---

*[docs/RFCS/distributed_sql.md, line 703 \[r3\]](https://reviewable.io:443/reviews/raduberinde/cockroach/7#-KFFaEMhh64lAVkunFoR:-KFFaEMhh64lAVkunFoS:-1489096328) ([raw file](https://github.com/raduberinde/cockroach/blob/d70937c07c782e58cbaeb8b0b059ca6baeb98c07/docs/RFCS/distributed_sql.md#L703)):*
remove the "probably", "seems like" etc. just say "in a couple of cases we don't want"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/raduberinde/cockroach/7#-:-KFFaX22KWMZuUX9zc5l:1417088721)*

**Comment by andreimatei** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[docs/RFCS/distributed_sql.md, line 692 \[r3\]](https://reviewable.io:443/reviews/raduberinde/cockroach/7#-KFFaQvo-PJ0kbw8ACrj:-KFFbGVgzQNF3QWs4Oq4:-499146009) ([raw file](https://github.com/raduberinde/cockroach/blob/d70937c07c782e58cbaeb8b0b059ca6baeb98c07/docs/RFCS/distributed_sql.md#L692)):*
Done.

---

*[docs/RFCS/distributed_sql.md, line 703 \[r3\]](https://reviewable.io:443/reviews/raduberinde/cockroach/7#-KFFaEMhh64lAVkunFoR:-KFFbGpVjL8pVHSxh6XU:-499146009) ([raw file](https://github.com/raduberinde/cockroach/blob/d70937c07c782e58cbaeb8b0b059ca6baeb98c07/docs/RFCS/distributed_sql.md#L703)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/raduberinde/cockroach/7)*

**Pull request merged**